Pull request for DB and transactor metrics

amaurySabran opened the pull request.

amaurySabran committed:
added metrics to track db connectionPool usage

amaurySabran committed:
added metrics to track the queries that take the most time to execute

amaurySabran committed:
modified the DbPoolManager and Transactor to use the metrics

amaurySabran committed:
made the transactor metrics package private

amaurySabran committed:
changed the mock test so that it can run with the transactor metrics

amaurySabran committed:
cleaned small differences with the master branch

amaurySabran committed:
modified:   jack-core/src/com/rapleaf/jack/transaction/DbMetrics.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/DbPoolManager.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java
	new file:   jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java
Added tests cases for the transactor and db metrics and remmoved irrelevant metric.

amaurySabran committed:
modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java
added a test case for the average Idle time

amaurySabran committed:
added a switch to enable or disable the metrics tracking

amaurySabran committed:
added a switch to enable or disable the metrics tracking

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,167 @@
+  private long totalActiveTime;
+  private long openedConnections;
+  //Transactor parameters
+
```
Can you unify the comment layout? In this case, it's better to keep one newline before the comment, remove the newline after the comment, and add one extra space between `//` and the comment.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,167 @@
+      }
+      currentConnections = newConnections;
+
+
```
Separation by one newline seems enough. There are some other places where there are two consecutive newlines. Can you unify them to just one? Sorry that I am kind of sensitive to code layout.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,167 @@
+    String summary = "";
+    summary += ("\n-----------------------TRANSACTOR METRICS-----------------------\n");
+    summary += ("\nAverage number of Idle connections is between " + String.format("%,.2f", getAverageIdleConnectionsMinValue()) + " and " + String
+        .format("%,.2f", getAverageIdleConnectionsMaxValue()));
```
This line was broken into two again. Probably need to further increase the line width limit in IntelliJ.

There are string concatenated to `String.format` call. In this case, we usually place the whole line into a `String.format`. That will make it easier to read.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbPoolManager.java:
```diff
@@ -71,6 +73,8 @@
 
     this.connectionPool = new GenericObjectPool<DB>(new DbPoolFactory<DB>(dbConstructor), config);
+    this.metrics = new DbMetricsImpl(config.getMaxTotal(), config.getMinIdle(), config
+        .getMaxWaitMillis(), config.getSoftMinEvictableIdleTimeMillis());
```
This line was also broken into two by IntelliJ.

amaurySabran committed:
Changed the tests so that they have wider acceptance values

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbPoolManager.java:
```diff
@@ -110,6 +120,33 @@ public void close() {
+    } else {
+      LOG.info("logging disabled, db metrics cannot be accessed");
+      return null;
+    }
```
Since the `metrics` is always constructed in the transactor, to avoid null pointer exception, you can still return `metrics` even when tracking is disabled. Just that all the method call from the return `metrics` will give zero, which is fine.

tuliren replied:
Another way is to create a second implementation of the metrics interface which does nothing. When tracking is disabled, assign the `metrics` variable to this implementation. But we probably don't need it for now.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/ITransactor.java:
```diff
@@ -20,6 +20,10 @@
+  boolean isMetricsTrackingEnabled();
+
+  void toggleMetricsTracking();
+
```
Let's not expose these options to users for now. I am thinking that users should be allowed to enable or disable tracking only when constructing the transactor. Can you add a method (e.g. `disableMetricsTracking`) in the builder of the transactor?

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,99 @@
+  synchronized void update(long executionTime, StackTraceElement queryStackTrace) {
+    if (longestQueries.containsKey(queryStackTrace)) {
+      if (longestQueries.get(queryStackTrace) < executionTime) {
+        longestTimes.removeFirstOccurrence(longestQueries.get(queryStackTrace));
```
`longestTimes` is a linked list. So the `removeFirstOccurrence` needs O(N) time, in which N is the list size. This could be inefficient if the list size is large. I know that for now the list size is always 10, so that's fine. But can you try to think of a better data structure?

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,99 @@
+  HashMap<StackTraceElement, Long> longestQueries;
+
+
+  //Contains a map with the *longestQueriesListSize* queries that took the most time to execute and their execution time
```
Better to give a brief description of the space-saving algorithm there. This is not a common algorithm.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,99 @@
+    String log = "";
+    log += "\n--------------" + longestQueriesListSize + " QUERIES WITH LONGEST EXECUTION TIME-----------------\n";
+    for (Map.Entry<StackTraceElement, Long> query : sortedMap.entrySet()) {
+      log += "\nClass name : " + query.getKey().getClassName();
```
`getClassName` will give a fully qualified name. A simple class name (just the last part of the full name) will be sufficient here. It will be much shorter, and make it possible to put everything in one line.

amaurySabran committed:
modified:   jack-core/src/com/rapleaf/jack/transaction/DbMetrics.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/DbPoolManager.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/IDbManager.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/ITransactor.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java
	modified:   jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestDbPoolManager.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestMockTransactorImpl.java
	modified:   jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java

Added the logging switch to the transactor builder

amaurySabran committed:
modified:   jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java

amaurySabran committed:
cleaned the testDbMetrics file

amaurySabran committed:
changed the tests so that travis can pass them

amaurySabran committed:
changed the metrics tests

amaurySabran committed:
removed file with ongoing experiments

amaurySabran committed:
changed the transactor metrics to track queries with large average time and track them efficiently

amaurySabran committed:
added query count to query metrics

amaurySabran committed:
changed a test that Travis struggled to pass

amaurySabran committed:
cleaned the typo of some files

amaurySabran committed:
changed the tests to actually measure the metrics and compare them to the ones measured by the DbMetrics and TransactorMetrics objects

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,67 @@
+  Comparator<TransactorMetricElement> queryComparator;
+
+  //Contains a map with the *longestQueriesListSize* queries that took the most time to execute and their execution time
+
```
I think some of the previously comments are ignored. Can you add a brief description of this space saving algorithm? This is not a common one, and could be confusing to others. You can add a link to the paper.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,160 @@
+  private static final Logger LOG = LoggerFactory.getLogger(DbMetricsImpl.class);
+
+  public DbMetricsImpl(int maxTotalConnections, int minIdleConnections, long maxWaitMillis, long
+      keepAliveMillis) {
```
Please keep all the arguments in one line.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,160 @@
+    summary += String.format("\nAverage number of Active connections : %,.2f", getAverageActiveConnections());
+    summary += ("\nTotal number of queries/executions : " + getTotalQueries());
+    summary += ("\nConnections opened : " + getOpenedConnectionsNumber());
+    summary += String.format("\n Max capacity time (%%)" + " : %,.2f ", getMaxConnectionsProportion());
```
The two strings should be merged into one string inside the `format` method.

tuliren replied:
Also please remove the extra space at the end of the string.

tuliren replied:
Should time the number from `getMaxConnectionsProportion()` with `100`, and append a `%` at the end to make it a percentage.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,160 @@
+
+    summary += String.format("\nAverage connection execution time : %,.2f ms", getAverageConnectionExecutionTime());
+    summary += String.format("\nAverage connection waiting time : %,.2f ms", getAverageConnectionWaitingTime());
+    summary += String.format("\nMaximum connection waiting time : %1$d    ms", getMaxConnectionWaitingTime());
```
Not sure why there are extra spaces before this `ms`. If you want to control the width of the output, you can use the width specifier in [Java Formatter](https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html).

Since the argument is not reused in this line, there is no need to use the index (`1$`) specifier.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,160 @@
+    summary += String.format("\nAverage connection execution time : %,.2f ms", getAverageConnectionExecutionTime());
+    summary += String.format("\nAverage connection waiting time : %,.2f ms", getAverageConnectionWaitingTime());
+    summary += String.format("\nMaximum connection waiting time : %1$d    ms", getMaxConnectionWaitingTime());
+    summary += ("\nTransactor lifetime : " + lifeTimeStopwatch.elapsedMillis() + "ms");
```
Missed a space before the `ms` here.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -108,6 +132,10 @@ private void execute(IExecution<DB> execution, boolean asTransaction) {
   public void close() {
+    if (metricsTrackingEnabled) {
+      LOG.info(dbManager.getMetrics().getSummary());
+      LOG.info("" + "\n" + (queryMetrics.getSummary()));
```
The first empty string should be deleted. The parentheses around the `queryMetrics.getSummary()` is also unnecessary.

When calling the `info` method on the logger, you can do something like this:
```java
LOG.info("\n{}", queryMetrics.getSummary());
```

Similar to `String.format`, later on `{}` will be replaced with the string argument specified by `queryMetrics.getSummary()`. The advantage of this approach is lazy evaluation. If the logger is configured to not log anything at the info level, the string concatenation will be omitted. This does not help much here, since the only thing being concatenated is a newline character. However, for complicated cases, it will help.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricElement.java:
```diff
@@ -0,0 +1,50 @@
+class TransactorMetricElementsComparator implements Comparator<TransactorMetricElement> {
+
+  public TransactorMetricElementsComparator() {
+  }
```
This constructor can be omitted.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricElement.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.Comparator;
+
+public class TransactorMetricElement {
```
Since this class is used as elements in collections, it is better to override the `hashCode` and `equals` methods. You can google about why they should be overridden and how to do that. You can also follow the implementations in `IndexHint`.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricElement.java:
```diff
@@ -0,0 +1,50 @@
+    return queryTrace;
+  }
+
+  public synchronized void addExecution(long executionTime) {
```
This method is only called in the `TransactorMetricsImpl#update`, which is already synchronized. So this method probably does not need to be synchronized.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,67 @@
+  TransactorMetricsImpl(int longestQueriesSize) {
+    this.longestQueriesSize = longestQueriesSize;
+    this.longestQueriesMap = new HashMap<>();
+    this.queryComparator = new TransactorMetricElementsComparator().reversed();
```
We are always using the same comparator, so the `reversed` order can be written into the comparator itself (switch the `1` and `-1` output). Also we can create a static variable in `TransactorMetricElement`, and reference that variable here. In this way, we can avoid creating new comparator objects.

tuliren replied:
A better approach is to make the `TransactorMetricElement` extend `Comparable`. In this way you don't need a separator comparator for the priority queue. In this way, you can define the `natural order` of `TransactorMetricElement` objects.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -25,8 +25,15 @@
   private final IDbManager<DB> dbManager;
 
-  TransactorImpl(IDbManager<DB> dbManager) {
+  private final int queryLogSize = 10;
```
Better to make this variable a static variable, because it does not change in different instances.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -65,14 +72,28 @@ public void executeAsTransaction(IExecution<DB> execution) {
         connection.commit();
       }
+      long executionTime = System.currentTimeMillis() - startTime;
+      queryMetrics.update(executionTime, Thread.currentThread().getStackTrace()[3]);
```
Should check `metricsTrackingEnabled` before updating queryMetrics.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java:
```diff
@@ -0,0 +1,67 @@
+    for (TransactorMetricElement query : longestQueriesList) {
+      log += "\nClass name : " + query.getQueryTrace().getClassName();
+      log += "\nLine number : " + query.getQueryTrace().getLineNumber();
+      log += String.format("\nAverage execution runtime : %,.2f", query.getAverageExecutionTime());
```
Please add a unit for the time.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java:
```diff
@@ -0,0 +1,200 @@
+
+  @After
+  public void cleanup() throws Exception {
+    executorService = null;
```
Executor service should be `shutdown` at the end. Setting it to `null` is not necessary.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java:
```diff
@@ -0,0 +1,200 @@
+    DbMetrics dbMetrics = transactor.getDbMetrics();
+    double maxConnectionsWaitingTime = dbMetrics.getMaxConnectionWaitingTime();
+    transactor.close();
+    double expectedMaxConnectionsWaitingTime = (finishingTime1 - startingTime2[0]);
```
The parentheses are unnecessary.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java:
```diff
@@ -0,0 +1,200 @@
+        () -> {
+          startingTime2[0] = stopwatch.elapsedMillis();
+          transactor.execute(a -> {sleepMillis(100);});
+        });
```
Can you unify the two lambda calls code format?

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java:
```diff
@@ -0,0 +1,200 @@
+    double averageIdleConnectionsMinValue = dbMetrics.getAverageIdleConnectionsMinValue();
+    transactor.close();
+
+    assert ((expectedAverageIdleConnections - .1 <= averageIdleConnectionsMaxValue) && (expectedAverageIdleConnections + .1 >= averageIdleConnectionsMinValue));
```
Better to use `assertEquals` method instead of `assert` keyword.

tuliren replied:
This line can actually be replaced with `assertRoughEqual`.

amaurySabran replied:
Actually one of the comparisons is with averageIdleConnectionsMinValue and the other with averageIdleConnectionsMaxValue so I can't use assertRoughEqual

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java:
```diff
@@ -0,0 +1,200 @@
+    assertRoughEqual(stopwatch.elapsedMillis() - startTime, dbMetrics.getLifeTime(), 20);
+  }
+
+  private void assertRoughEqual(double value, double expected, double error) {
```
This method can probably be replaced with [`assertEquals` for double](http://junit.org/junit4/javadoc/latest/src-html/org/junit/Assert.html#line.550), which accepts a `delta` argument that is similar to `error` here.

tuliren replied:
Never mind. Some of the values passed into this method are `long`s. So using `assertEquals` will need type conversion, which is not very convenient.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java:
```diff
@@ -0,0 +1,73 @@
+  private void assertRoughEqual(double value, double expected, double error) {
+    assertTrue(value <= (expected + error) && value >= (expected - error));
+  }
+}
```
Please add a newline at the end. Here is the reference for why: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java:
```diff
@@ -104,7 +103,7 @@ public void testTransactionRollbackSqlException() throws Exception {
   }
 
-  @Test(timeout = 3 * 1000) // 3s
+  @Test(timeout = 5 * 1000) // 5s
```
Did this test time out on Travis?

amaurySabran replied:
yes it did

tuliren replied:
Got it. Thanks for the fix.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java:
```diff
@@ -0,0 +1,73 @@
+
+  @After
+  public void cleanup() throws Exception {
+    executorService = null;
```
Again, executor service needs to be `shutdown`, and setting it to `null` is unnecessary.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java:
```diff
@@ -0,0 +1,73 @@
+    int lastLine = 0;
+    transactor.close();
+    for (TransactorMetricElement query : queryMetrics.getLongestQueries()) {
+      assert (query.getQueryTrace().getLineNumber() > lastLine);
```
Please use `assertTrue` instead of `assert` keyword.

tuliren commented on jack-test/test/java/com/rapleaf/jack/transaction/TestTransactorMetrics.java:
```diff
@@ -0,0 +1,73 @@
+    assertRoughEqual(executionTime, maxExecutionTime, 20);
+  }
+
+  private void assertRoughEqual(double value, double expected, double error) {
```
This is duplicated in two tests. You can move it to `JackTestCase`.

amaurySabran replied:
One of them is for long values and the other for double values. Should I just keep the double one and move it to `JackTestCase`?

tuliren replied:
Well, if each of them is just used in one file, it's fine to keep them as is.

tuliren commented:
@amaurySabran, very good! I really like many of the test cases you have written.

amaurySabran committed:
fixed some errors with the code and added equal method to the TransactorMetricElement object

amaurySabran committed:
fixed a problem with the executorService cleanup in test files

amaurySabran committed:
added new line at the end of file

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java:
```diff
@@ -0,0 +1,159 @@
+    summary += String.format("\nAverage number of Active connections : %,.2f", getAverageActiveConnections());
+    summary += ("\nTotal number of queries/executions : " + getTotalQueries());
+    summary += ("\nConnections opened : " + getOpenedConnectionsNumber());
+    summary += String.format("\n Max capacity time (%%) : %,.2f %%", 100 * getMaxConnectionsProportion());
```
There is an extra space before "max". Also the "(%%)" can be omitted, since it is already added after the number.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricElement.java:
```diff
@@ -0,0 +1,68 @@
+    if (this == obj) {
+      return true;
+    }
+    if (!(this.getClass() == obj.getClass())) {
```
"obj" could be null, and this check could throw null pointer exception. Should use "instanceof", which will return false if obj is null.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorMetricElement.java:
```diff
@@ -0,0 +1,68 @@
+      return false;
+    } else {
+      TransactorMetricElement that = (TransactorMetricElement)obj;
+      return ((this.queryTrace.equals(that.queryTrace)) && (this.totalExecutionTime == that.totalExecutionTime) && (this.count == that.count));
```
The extra parentheses are unnecessary.

tuliren commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -25,8 +25,15 @@
   private final IDbManager<DB> dbManager;
 
-  TransactorImpl(IDbManager<DB> dbManager) {
+  private static final int queryLogSize = 10;
```
Change the name of variableto upper camel case. This is the naming convention for static fields.

Also bump it to a larger number, like 30.

tuliren reviewed (commented):
Just a few comments. Have you tested the overhead of query tracking? Should make sure the tracking does not significantly impact query performance.

After answering the few comments, you can close this pull request and open a new one against the liveramp jack repo.

Nice job!

amaurySabran committed:
added a test to check that the query tracking doesnt add serious query overhead

amaurySabran committed:
changed query overhead test to make it more reliable

amaurySabran committed:
fixed a small bug in the query overhead test

amaurySabran committed:
cleaned the last test case

amaurySabran committed:
moved the average query execution time measurement from dbmetrics to transactor metrics

amaurySabran committed:
tried to fix a test bug on travis
